File: plugins-community/ascendc-ops-lab-developer/workflows/templates/archive_tasks/gather_elements_v2/kernel/op_host/gather_elements_v2.cpp
/*
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include "torch_kernel_helper.h"
#include "tiling/platform/platform_ascendc.h"
#include "aclrtlaunch_gather_elements_v2_last_dim_fp16.h"
#include "aclrtlaunch_gather_elements_v2_last_dim_fp32.h"
#include "aclrtlaunch_gather_elements_v2_transpose_fp16.h"
#include "aclrtlaunch_gather_elements_v2_transpose_fp32.h"
#include "aclrtlaunch_gather_elements_v2_scalar_fp16.h"
#include "aclrtlaunch_gather_elements_v2_scalar_fp32.h"
#include "gather_elements_v2_tiling.h"

namespace ascend_kernel {

constexpr int32_t GATHER_MODE_LAST_DIM = 0;
constexpr int32_t GATHER_MODE_TRANSPOSE = 1;
constexpr int32_t GATHER_MODE_SCALAR = 2;

inline int32_t CeilDivI32(int32_t a, int32_t b)
{
    if (b == 0) { return 0; }
    return (a + b - 1) / b;
}

inline int32_t ChooseBlockM(int32_t m)
{
    int32_t blockM = 32;
    if (m < 32) blockM = 16;
    if (m < 16) blockM = 8;
    if (m < 8) blockM = 4;
    if (m < 4) blockM = 2;
    if (m < 2) blockM = 1;
    return blockM;
}

static inline void LaunchGatherElementsV2Kernel(
    const at::Tensor &x, const at::Tensor &index, const at::Tensor &row_map,
    at::Tensor &y, int32_t xStride, int32_t yStride, int32_t ig32,
    int32_t blockM, int32_t usedCoreNum, int32_t tasksPerCore,
    int32_t subBlockM, int32_t m, int32_t xRows, int32_t mode32)
{
    at::Tensor tilingCpu = at::empty(
        {static_cast<int64_t>(sizeof(GatherElementsV2KernelTiling))},
        at::device(at::kCPU).dtype(at::kByte));
    auto *tiling = reinterpret_cast<GatherElementsV2KernelTiling *>(tilingCpu.data_ptr());
    tiling->M = m;                tiling->XRows = xRows;
    tiling->XG = xStride;        tiling->IG = ig32;
    tiling->XStride = xStride;   tiling->YStride = yStride;
    tiling->blockM = blockM;     tiling->usedCoreNum = usedCoreNum;
    tiling->tasksPerCore = tasksPerCore;
    tiling->subBlockM = subBlockM;
    tiling->useRowMap = (mode32 == GATHER_MODE_LAST_DIM) ? 0 : 1;
    tiling->mode = mode32;

    auto tilingNpu = tilingCpu.to(at::kPrivateUse1);
    uint32_t blockDim = static_cast<uint32_t>(usedCoreNum);

    bool isFp16 = (x.scalar_type() == at::kHalf);
    if (mode32 == GATHER_MODE_LAST_DIM) {
        if (isFp16) {
            EXEC_KERNEL_CMD(gather_elements_v2_last_dim_fp16, blockDim, x, index, row_map, y, tilingNpu);
        } else {
            EXEC_KERNEL_CMD(gather_elements_v2_last_dim_fp32, blockDim, x, index, row_map, y, tilingNpu);
        }
    } else if (mode32 == GATHER_MODE_TRANSPOSE) {
        if (isFp16) {
            EXEC_KERNEL_CMD(gather_elements_v2_transpose_fp16, blockDim, x, index, row_map, y, tilingNpu);
        } else {
            EXEC_KERNEL_CMD(gather_elements_v2_transpose_fp32, blockDim, x, index, row_map, y, tilingNpu);
        }
    } else {
        if (isFp16) {
            EXEC_KERNEL_CMD(gather_elements_v2_scalar_fp16, blockDim, x, index, row_map, y, tilingNpu);
        } else {
            EXEC_KERNEL_CMD(gather_elements_v2_scalar_fp32, blockDim, x, index, row_map, y, tilingNpu);
        }
    }
}

at::Tensor gather_elements_v2(
    const at::Tensor &x,
    const at::Tensor &index,
    const at::Tensor &row_map,
    int64_t ig,
    int64_t mode)
{
    TORCH_CHECK(x.device().is_privateuseone(), "x must be on NPU");
    TORCH_CHECK(index.device().is_privateuseone(), "index must be on NPU");
    TORCH_CHECK(row_map.device().is_privateuseone(), "row_map must be on NPU");
    TORCH_CHECK(x.dim() == 2, "x must be [XRows, XStride]");
    TORCH_CHECK(index.dim() == 2, "index must be [M, YStride]");
    TORCH_CHECK(row_map.dim() == 1, "row_map must be [M]");
    TORCH_CHECK(x.is_contiguous(), "x must be contiguous");
    TORCH_CHECK(index.is_contiguous(), "index must be contiguous");
    TORCH_CHECK(row_map.is_contiguous(), "row_map must be contiguous");
    TORCH_CHECK(index.scalar_type() == at::kInt, "index must be int32");
    TORCH_CHECK(row_map.scalar_type() == at::kInt, "row_map must be int32");
    TORCH_CHECK(x.scalar_type() == at::kHalf || x.scalar_type() == at::kFloat, "x must be float16 or float32");

    const int32_t mode32 = static_cast<int32_t>(mode);
    const int32_t m = static_cast<int32_t>(index.size(0));
    TORCH_CHECK(index.size(0) == row_map.size(0), "index and row_map row count must match");
    const int32_t xRows = static_cast<int32_t>(x.size(0));
    const int32_t xStride = static_cast<int32_t>(x.size(1));
    const int32_t yStride = static_cast<int32_t>(index.size(1));
    const int32_t ig32 = static_cast<int32_t>(ig);

    TORCH_CHECK(ig32 >= 0, "ig must be non-negative");
    TORCH_CHECK(ig32 <= yStride, "ig must not exceed index width");

    const int32_t blockM = ChooseBlockM(std::max(m, 1));
    const int32_t mNum = CeilDivI32(std::max(m, 1), blockM);
    const int32_t usedCoreNum = std::min(DEFAULT_NUM_PHYSICAL_CORES, std::max(mNum, 1));
    const int32_t tasksPerCore = CeilDivI32(mNum, usedCoreNum);
    const int32_t subBlockM = std::max(blockM / DEFAULT_VEC_NUM, 1);

    at::Tensor y = at::zeros({m, yStride}, x.options());
    LaunchGatherElementsV2Kernel(x, index, row_map, y, xStride, yStride, ig32, blockM,
                                  usedCoreNum, tasksPerCore, subBlockM, m, xRows, mode32);

    return y;
}

} // namespace ascend_kernel

File: plugins-community/tilelang2ascendc-ops-generator/workflows/templates/archive_tasks/gather_elements_v2/design/tile_level/gather_elements_v2.py
#!/usr/bin/env python3
# ----------------------------------------------------------------------------------------------------------
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ----------------------------------------------------------------------------------------------------------

import sys
from dataclasses import dataclass
from pathlib import Path

import tilelang
import tilelang.language as T

_COMMON_DIR = Path(__file__).resolve().parent.parent
if str(_COMMON_DIR) not in sys.path:
    sys.path.insert(0, str(_COMMON_DIR))

from gather_elements_v2_common import (  
    pass_configs,
    _kernel_setup,
    _gather_cols,
    RowProcessCtx,
    _process_rows_for_bx,
)


# NOTE: The two @T.prim_func functions below (last_dim_kernel, indexed_kernel)
# share the same pipeline boilerplate (Kernel setup, UB allocation, gather
# skeleton) except for the RowMap indirection in indexed_kernel.  The common
# helpers (_gather_cols, _process_rows_for_bx) are defined in
# gather_elements_v2_common and called from each prim_func.


@dataclass
class GatherConfig:
    m_size: int
    x_rows: int
    x_g: int
    i_g: int
    x_stride: int
    y_stride: int
    mode: str = "last_dim"
    dtype: str = "float32"


@tilelang.jit(out_idx=[2], pass_configs=pass_configs)
def gather_elements_v2(cfg):
    m_size, x_rows, x_g, i_g, x_stride, y_stride, mode, dtype = (
        cfg.m_size, cfg.x_rows, cfg.x_g, cfg.i_g, cfg.x_stride, cfg.y_stride, cfg.mode, cfg.dtype
    )
    c = _kernel_setup(m_size)
    block_m, m_num, used_core_num = c.block_m, c.m_num, c.used_core_num
    tasks_per_core, sub_block_m = c.tasks_per_core, c.sub_block_m

    def _gather_scope_body(*args):
        cid, vid, x_ub, index_ub, out_ub, x_in, index_in, y_out, row_map = args
        for local_idx in T.serial(tasks_per_core):
            bx = cid * tasks_per_core + local_idx
            if bx < m_num:
                _process_rows_for_bx(RowProcessCtx(
                    bx, vid, sub_block_m, block_m, m_size,
                    x_in, index_in, y_out, i_g,
                    x_ub, index_ub, out_ub, row_map=row_map,
                ))

    @T.prim_func
    def last_dim_kernel(
        x_in: T.Tensor((m_size, x_stride), dtype),
        index_in: T.Tensor((m_size, y_stride), "int32"),
        y_out: T.Tensor((m_size, y_stride), dtype),
    ):
        with T.Kernel(used_core_num, is_npu=True) as (cid, vid):
            x_ub = T.alloc_ub((1, x_stride), dtype)
            index_ub = T.alloc_ub((1, y_stride), "int32")
            out_ub = T.alloc_ub((1, y_stride), dtype)

            with T.Scope("V"):
                _gather_scope_body(cid, vid, x_ub, index_ub, out_ub,
                                   x_in, index_in, y_out, None)

    @T.prim_func
    def indexed_kernel(
        x_in: T.Tensor((x_rows, x_stride), dtype),
        index_in: T.Tensor((m_size, y_stride), "int32"),
        row_map: T.Tensor((m_size,), "int32"),
        y_out: T.Tensor((m_size, y_stride), dtype),
    ):
        with T.Kernel(used_core_num, is_npu=True) as (cid, vid):
            x_ub = T.alloc_ub((1, x_stride), dtype)
            index_ub = T.alloc_ub((1, y_stride), "int32")
            out_ub = T.alloc_ub((1, y_stride), dtype)

            with T.Scope("V"):
                _gather_scope_body(cid, vid, x_ub, index_ub, out_ub,
                                   x_in, index_in, y_out, row_map)

    if mode == "last_dim":
        return last_dim_kernel
    # transpose mode uses the same indexed kernel logic
    return indexed_kernel
